ui/apps/platform/src/services/IntegrationsService.ts

```diff
@@ -103,11 +106,20 @@ export function saveIntegrationV2(
         const config = data[getJsonFieldBySource(source)] as IntegrationBase;
         return axios.patch(`${getPath(source)}/${config.id}`, data);
     }
-    // Machine access configs should be wrapped.
-    if (source === 'authProviders') {
-        return updateMachineAccessConfig(data as AuthMachineToMachineConfig);
+    // Some services expect requests to be wrapped in dedicated proto messages.
+    // In these cases, the top level request payload does not contain the integration id.
+    // Taking cloud source requests as an example, the integration id is accessed as
+    // `data.cloudSource.id` rather than `data.id`.
+    // Additionally, services may differ in whether `put` or `patch` is used to update
+    // existing integrations.
+    switch (source) {
+        case 'authProviders':
+            return updateMachineAccessConfig(data as AuthMachineToMachineConfig);
+        case 'cloudSources':
+            return updateCloudSource(data as UpdateCloudSourceRequest);
```

Comment:
Stephan, The additions make sense. Here is where I got stuck.

Does `saveIntegration` function also need special cases, if `updatePassword: true` according to here:
https://github.com/stackrox/stackrox/blob/master/ui/apps/platform/src/Containers/Integrations/hooks/useIntegrationActions.ts#L46-L50

Reply:
Some background:

I think the use of `updatePassword` as a heuristic is flawed, because it makes assumptions about the API that are not true. We have some APIs that take the schema `{config: data}` vs `{data}` - but this cannot be mapped to whether the update API takes `updatePassword` (e.g. cloud sources where it's called `updateCredentials` instead). There is another assumption that if `updatePassword` exists, then the create API is unwrapped (see https://github.com/stackrox/stackrox/blob/2c3323ce02da5d47f2702f91f802ba1973a1b545/ui/apps/platform/src/services/IntegrationsService.ts#L122).

Note that the new API guidelines recommend wrapping request data in separate proto messages (https://github.com/stackrox/stackrox/blob/master/.github/api_guidelines.md). But not all of our APIs follow this convention yet, nor will they any time soon. So I think we should get rid of the `updatePassword` checks, because it's an assumption that's sometimes wrong and is not part of the API contract.

Now having said that, what special case do you mean? What I tried to do here is that for cloud sources, we simply decide the schema based on the concrete type (`case 'cloudSources'`). I still left the `updatePassword` check in there to keep the diff small, but as pointed out above I think those should be removed entirely. For cloud sources we have `hasUpdatePassword==false`, which means we take `saveIntegrationV2` over `saveIntegration` and drop to the type case below.

Reply:
Aha, `updateCredentials` property! Now I understand. Thank you, Stephan.